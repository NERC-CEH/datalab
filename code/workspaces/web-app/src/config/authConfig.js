import axios from 'axios';
import Oidc from 'oidc-client';
import Promise from 'bluebird';
import Auth from '../auth/auth';

let authSession;

async function getAuthConfig() {
  const { data } = await axios.get('/web_auth_config.json');

  // Overwrite default configuration for running locally/development purposes
  return window.location.hostname.match(/localhost/) ? localData(data) : data;
}

function localData(data) {
  const newData = data;
  newData.oidc.userManager.silent_redirect_uri = `${window.location.origin}/silent_callback`;
  newData.oidc.userManager.redirect_uri = `${window.location.origin}/callback`;
  return newData;
}

async function initialiseAuth() {
  const authConfig = await getAuthConfig();
  Oidc.Log.logger = console;
  Oidc.Log.level = Oidc.Log.INFO;

  const userManagerConfig = {
    ...authConfig.oidc.userManager,
    userStore: new Oidc.WebStorageStateStore(),
  };

  const userManager = new Oidc.UserManager(userManagerConfig);
  const PromisifyUserManager = Promise.promisifyAll(userManager);

  authSession = new Auth(userManager, PromisifyUserManager, authConfig);
}

const getAuth = () => (authSession);

export { initialiseAuth, getAuth };
