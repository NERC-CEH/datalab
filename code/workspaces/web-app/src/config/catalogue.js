import axios from 'axios';

export async function catalogueConfig() {
  const { data } = await axios.get('/catalogue_config.json');
  return data;
}

export async function catalogueAvailable() {
  const config = await catalogueConfig();
  return config.available;
}
