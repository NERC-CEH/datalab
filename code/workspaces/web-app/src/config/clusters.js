import axios from 'axios';

// This is a cached value.
// Initialise asynchronously before React starts.
// Get synchronously within React.
let cachedData;

export async function initialiseClusters() {
  const { data } = await axios.get('/clusters_config.json');
  cachedData = data;
}

export const getClusterMaxWorkers = type => cachedData[type.toLowerCase()].cluster.workersMax;
export const getWorkerMemoryMax = type => cachedData[type.toLowerCase()].workers.memoryMax_GB;
export const getWorkerCpuMax = type => cachedData[type.toLowerCase()].workers.CpuMax_vCPU;
export const getCondaRequired = type => cachedData[type.toLowerCase()].condaRequired;
