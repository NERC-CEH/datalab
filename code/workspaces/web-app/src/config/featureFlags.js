import axios from 'axios';

// This is a cached value.
// Initialise asynchronously before React starts.
// Get synchronously within React.
let featureFlags = {
  requestProjects: false,
};

async function initialiseFeatureFlags() {
  const { data } = await axios.get('/feature_flags_config.json');
  featureFlags = data;
}

const getFeatureFlags = () => (featureFlags);

export { initialiseFeatureFlags, getFeatureFlags };
