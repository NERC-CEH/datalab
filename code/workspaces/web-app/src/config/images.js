import axios from 'axios';

const NOTEBOOK_CATEGORY = 'ANALYSIS';
const SITE_CATEGORY = 'PUBLISH';

// These are cached values.
// Initialise asynchronously before React starts.
// Get synchronously within React.
let cachedData;
let cachedNotebooks;
let cachedSites;
let cachedUserActions;

async function initialiseImages() {
  const { data } = await axios.get('/image_config.json');
  cachedData = data;
  cachedNotebooks = getInfoForItemsInCategory(NOTEBOOK_CATEGORY);
  cachedSites = getInfoForItemsInCategory(SITE_CATEGORY);
  cachedUserActions = getInfoForUserActions();
}

function getInfoForItemsInCategory(category) {
  return Object.entries(cachedData.types).reduce(
    (infoForType, [key, value]) => {
      if (value.category === category) {
        infoForType[key] = value; // eslint-disable-line no-param-reassign
      }
      return infoForType;
    },
    {},
  );
}

function getInfoForUserActions() {
  return Object.entries(cachedData.types).reduce(
    (infoForType, [key, value]) => {
      infoForType[key] = calculateUserActionsForType(key); // eslint-disable-line no-param-reassign
      return infoForType;
    },
    {},
  );
}

function calculateUserActionsForType(type) {
  const typeInfo = cachedData.types[type];

  const actionDefaults = getActionDefaults();
  let categoryDefaultActions = {};
  let actionOverrides = {};

  if (typeInfo) {
    categoryDefaultActions = cachedData.categories[typeInfo.category].userActions || {};
    actionOverrides = typeInfo.userActionOverrides || {};
  }

  return { ...actionDefaults, ...categoryDefaultActions, ...actionOverrides };
}

function getActionDefaults() {
  const possibleActions = ['share', 'edit', 'restart', 'delete', 'logs'];
  const defaultValue = true;
  return possibleActions.reduce(
    (returnValue, actionName) => {
      returnValue[actionName] = defaultValue; // eslint-disable-line no-param-reassign
      return returnValue;
    }, {},
  );
}

const getNotebookInfo = () => (cachedNotebooks);
const getSiteInfo = () => (cachedSites);
const getUserActionsForType = type => (cachedUserActions[type] || getActionDefaults());

export { initialiseImages, getNotebookInfo, getSiteInfo, getUserActionsForType };
