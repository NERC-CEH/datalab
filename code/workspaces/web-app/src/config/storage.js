import axios from 'axios';

let cachedData;

async function storageConfig() {
  if (cachedData) {
    return cachedData;
  }

  const { data } = await axios.get('/storage_config.json');
  cachedData = data;
  return cachedData;
}

export async function storageTypes() {
  const data = await storageConfig();
  return Object.keys(data.types);
}

export async function storageCreationDefaultType() {
  const data = await storageConfig();
  return data.creationOptions.defaultType;
}

export async function storageCreationAllowedDisplayOptions() {
  const data = await storageConfig();
  return data.creationOptions.allowedTypes
    .map(type => ({
      text: data.types[type].displayValue,
      value: type,
    }));
}

export async function storageDisplayValue(type) {
  const data = await storageConfig();
  return data.types[type].displayValue;
}

export async function storageDescription(type) {
  const data = await storageConfig();
  return data.types[type].description;
}
