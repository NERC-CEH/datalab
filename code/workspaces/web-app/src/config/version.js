import axios from 'axios';

// This is a cached value.
// Initialise asynchronously before React starts.
// Get synchronously within React.
let version;

async function initialiseVersion() {
  const { data } = await axios.get('/version.json');
  ({ version } = data);
}

const getVersion = () => (version);

export { initialiseVersion, getVersion };
