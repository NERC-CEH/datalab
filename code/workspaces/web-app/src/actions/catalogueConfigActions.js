import { catalogueConfig } from '../config/catalogue';

export const LOAD_CATALOGUE_CONFIG_ACTION = 'LOAD_CATALOGUE_CONFIG_ACTION';

const loadCatalogueConfig = () => ({
  type: LOAD_CATALOGUE_CONFIG_ACTION,
  payload: catalogueConfig(),
});

export default {
  loadCatalogueConfig,
};
