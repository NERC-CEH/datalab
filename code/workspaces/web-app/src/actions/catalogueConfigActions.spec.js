import { catalogueConfig } from '../config/catalogue';
import catalogueConfigActions from './catalogueConfigActions';

jest.mock('../config/catalogue');

describe('loadCatalogConfig', () => {
  it('created action matches snapshot', () => {
    catalogueConfig.mockReturnValueOnce('expected catalogue config value');
    expect(catalogueConfigActions.loadCatalogueConfig()).toMatchSnapshot();
  });
});
