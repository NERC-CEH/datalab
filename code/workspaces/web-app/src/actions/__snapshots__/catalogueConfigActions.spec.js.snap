// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadCatalogConfig created action matches snapshot 1`] = `
Object {
  "payload": "expected catalogue config value",
  "type": "LOAD_CATALOGUE_CONFIG_ACTION",
}
`;
